dll/cpl/sysdm/environment.c

```diff
@@ -17,6 +23,140 @@ typedef struct _VARIABLE_DATA
+    LPTSTR lpTemp;
+    LPTSTR lpToken;
+
+    dwValueLength = _tcslen(lpRawValue) + 1;
+    lpTemp = GlobalAlloc(GPTR, dwValueLength * sizeof(TCHAR));
+
+    StringCchCopy(lpTemp, dwValueLength, lpRawValue);
+    lpToken = _tcstok(lpTemp, _T(";"));
+
+    while (lpToken != NULL)
+    {
+        /* If the string has environment variable then expand it */
+        ExpandEnvironmentStrings(lpToken, lpToken, dwValueLength);
+
+        if (!PathIsDirectoryW(lpToken))
+            return FALSE;
+
+        lpToken = _tcstok(NULL, _T(";"));
+    }
+    GlobalFree(lpTemp);
+
+    return TRUE;
+}
+
+static DWORD
+GatherDataFromEditBox(HWND hwndDlg,
+                      PVARIABLE_DATA VarData)
+{
+    DWORD dwNameLength;
+    DWORD dwValueLength;
+
+    dwNameLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_NAME, WM_GETTEXTLENGTH, 0, 0);
+    dwValueLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_VALUE, WM_GETTEXTLENGTH, 0, 0);
+
+    if (dwNameLength == 0 || dwValueLength == 0)
+    {
+        return 0;
+    }
+
+    /* Reallocate the name buffer, regrowing it if necessary */
+    if (!VarData->lpName || (_tcslen(VarData->lpName) < dwNameLength))
+    {
+        if (VarData->lpName)
+            GlobalFree(VarData->lpName);
+        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
+    SendDlgItemMessage(hwndDlg, IDC_VARIABLE_NAME, WM_GETTEXT, dwNameLength + 1, (LPARAM)VarData->lpName);
+
+    /* Reallocate the name buffer, regrowing it if necessary */
+    if (!VarData->lpRawValue || (_tcslen(VarData->lpRawValue) < dwValueLength))
+    {
+        if (VarData->lpRawValue)
+            GlobalFree(VarData->lpRawValue);
+        VarData->lpRawValue = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
+    SendDlgItemMessage(hwndDlg, IDC_VARIABLE_VALUE, WM_GETTEXT, dwValueLength + 1, (LPARAM)VarData->lpRawValue);
+
+    return dwValueLength;
+}
+
+static DWORD
+GatherDataFromListView(HWND hwndListView,
+                       PVARIABLE_DATA VarData)
+{
+    DWORD dwValueLength;
+    DWORD NumberOfItems;
+    DWORD i;
+    TCHAR szData[MAX_PATH];
+
+    /* Gather the number of items for the semi-colon */
+    NumberOfItems = ListView_GetItemCount(hwndListView);
+
+    if (NumberOfItems == 0)
+    {
+        return 0;
+    }
+
+    /* Since the last item doesn't need the semi-colon subtract 1 */
+    dwValueLength = NumberOfItems - 1;
+
+    for (i = 0; i < NumberOfItems; i++)
+    {
+        ListView_GetItemText(hwndListView,
+                             i,
+                             0,
+                             szData,
+                             _countof(szData));
+        dwValueLength += _tcslen(szData);
+    }
+
+    if (VarData->lpRawValue == NULL)
+    {
+        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    }
+    else
+    {
+        GlobalFree(VarData->lpRawValue);
+        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    }
```

Comment:
Suggestion to regrow only if the new calculated size is larger than the old one:
```suggestion
    /* Reallocate the value buffer, regrowing it if necessary */
    if (!VarData->lpRawValue || (_tcslen(VarData->lpRawValue) < dwValueLength))
    {
        if (VarData->lpRawValue)
            GlobalFree(VarData->lpRawValue);
        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
    }
```
Otherwise, the current code (with always recreation of the buffer) would simplify to:
```suggestion
    /* Reallocate the value buffer */
    if (VarData->lpRawValue)
        GlobalFree(VarData->lpRawValue);
    VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
```

Reply:
@HBelusca Chose the first option.